### ISSUE TITLE
Multi-Label classification - Loading saved model do not reproduce same results 

### ISSUE THREAD

**Issue by thiagosantos1** [ThilinaRajapakse/simpletransformers#713]
**Describe the bug**
When running/training for multi-label classification(ClassificationModel), I save every epoch of the model. After training is done, with the current model I have I can then test the model and it works fine(let's say f1 of 98%). When loading the model and give the exact same test data, in the exact same order, it does not reproduce the results( f1 goes to 40% for instance --> just shooting most of the time to only 1 class). My guess is: Load model is either loading early epochs or not loading the learned weights. I did try to load every epoch folder, one by one, but got same behavior.

**To Reproduce**
I am working with 5 classes and the model when training is as follow:
model = ClassificationModel('bert', 'bert-base-uncased', num_labels=5, use_cuda=False,
    args={'learning_rate':1e-4, 'num_train_epochs': 50, 'reprocess_input_data': False, 
          'overwrite_output_dir': True, 'output_dir':'models/'})
Training and testing at the same instance I get F1 of 98% for instance

When testing(execute the program again):
model = ClassificationModel('bert', 'output_dir':'models/', num_labels=5, use_cuda=False,
    args={'learning_rate':1e-4, 'num_train_epochs': 50, 'reprocess_input_data': False, 
          'overwrite_output_dir': True, 'output_dir':self.'models/tuned/'})

I get F1 of < 60%. Shooting for only 1 class most of the time(around 95% of the test cases)

**Expected behavior**
Results of testing on the same data should be consistent, at all times.

**Desktop (please complete the following information):**
 - MACOS

**Additional context**
When I create the model I set to overwrite_output_dir:True. 
Maybe this affects in some how? Since I train the model(which saves) but maybe when I execute test on the same running instance, it overwrites again? Not sure what's going on

Am I using the wrong model for multlabel classification?

**Comment by thiagosantos1** [comment]
Any updates ? I have tried all possible things. It does seems like to be a bug

**Comment by BanZiXY** [comment]
I have encountered the same problem

**Comment by ThilinaRajapakse** [comment]
I can't reproduce this. I tested with the Kaggle toxic comments dataset.

Train:

```python
from simpletransformers.classification import MultiLabelClassificationModel
import pandas as pd
from sklearn.model_selection import train_test_split


train_df = pd.read_csv('data/train.csv')

train_df['labels'] = list(zip(train_df.toxic.tolist(), train_df.severe_toxic.tolist(), train_df.obscene.tolist(), train_df.threat.tolist(),  train_df.insult.tolist(), train_df.identity_hate.tolist()))
train_df['text'] = train_df['comment_text'].apply(lambda x: x.replace('\n', ' '))
train_df.to_csv('data/train_current.csv', index=False)

train_df, eval_df = train_test_split(train_df, test_size=0.2)

eval_df.to_csv("data/val.tsv", sep="\t")

print(train_df.head())

train_args = {
    'reprocess_input_data': True,
    'train_batch_size': 32,
    'gradient_accumulation_steps':8,
    'learning_rate': 3e-5,
    'num_train_epochs': 1,
    'max_seq_length': 128,
    'save_steps': -1,
    'overwrite_output_dir': True,
    'threshold': 0.5,
    'fp16': True,
    "evaluate_during_training": int(len(train_df) / 5),
    "wandb_project": "Multi Label Test"
}

model = MultiLabelClassificationModel('roberta', 'roberta-base', num_labels=6, args=train_args)

model.train_model(train_df, eval_df=eval_df)

```

Evaluation:

```python
from simpletransformers.classification import MultiLabelClassificationArgs, MultiLabelClassificationModel
import pandas as pd


train_args = {
    'eval_batch_size': 64,
}

eval_df = pd.read_csv("data/val.tsv", sep="\t")
eval_df["labels"] = eval_df["labels"].apply(lambda x: list(eval(x)))

print(eval_df)

model = MultiLabelClassificationModel("roberta", "outputs", args=train_args)

result, _, _ = model.eval_model(eval_df)
print(result)

```

Both get 0.99+ LRAP score. Am I missing something?

@BanZiXY Are you also on Mac?

**Comment by thiagosantos1** [comment]
@ThilinaRajapakse The problem is when using Multi-Class Classification. I am not sure if the error also happens for Multi label classification. 
Could you try to replicate by using  Multi-Class Classification ? Something like:

train_df = pd.DataFrame(train_data)
eval_df = pd.DataFrame(eval_data)
test_df = pd.DataFrame(test_data)
model_type = "bert"
model_name = "bert-base-cased"
train_args = {
    "reprocess_input_data": True,
    "overwrite_output_dir": True,
    "use_cached_eval_features": True,
    "output_dir": f"outputs/{model_type}",
    "best_model_dir": f"outputs/{model_type}/best_model",
    "evaluate_during_training": True,
    "max_seq_length": 128,
    "num_train_epochs": 10,
    "evaluate_during_training_steps": 1000,
    "wandb_project": "Classification Model Comparison",
    "wandb_kwargs": {"name": model_name},
    "save_model_every_epoch": False,
    "save_eval_checkpoints": False,
    "train_batch_size": 128,
    "eval_batch_size": 64,
}

train = False

if train:
  model = ClassificationModel(model_type, model_name, args=train_args,use_cuda=False,num_labels=3)
  model.train_model(train_df,eval_df=eval_df)
  result, _, _ = model.eval_model(eval_df,f1=f1_multiclass, acc=accuracy_score)
  print(result)

else:
  model = ClassificationModel(model_type, f"outputs/{model_type}", use_cuda=False, args=train_args,num_labels=3)
  print("\n\tEvaluating model")
  results, _, _ = model.eval_model(test_df, f1=f1_multiclass, acc=accuracy_score)
  print("\tResults\n\n", results)

**Comment by BanZiXY** [comment]
This happens when I load the local model

**Comment by thiagosantos1** [comment]
Seems like the problem is on the arguments. 
The definition of loading a model is as follow:
"To load a saved model, provide the path to the directory containing the saved model as the model_name. Note that you will need to specify the correct (usually the same used in training) args when loading the model"

Thus, I am using the same args dictionary for when creating and loading the model. If I do not give the same arguments, if only giving the 'eval_batch_size' when loading, then the model works just a good.

**Comment by AbinayaM02** [comment]
@thiagosantos1 : When you train a model, model_args.json file is automatically saved in the best_model_dir you specify. So, while loading the model, you can also load this model_args.json file and pass it to the args parameter. Your results should be reproducible now.

**Comment by thiagosantos1** [comment]
@AbinayaM02 How about to re-train a model ? When loading the model, can I continue the training ? Does it save the new update model to the same directory it was loaded?

**Comment by AbinayaM02** [comment]
Yes, it will save the new model in the same directory if you choose to overwrite it. You'll have to set overwrite_output_dir to True.

**Comment by ninasujit2016** [comment]
Sorry that i still got the exactly same problem as @thiagosantos1 mentioned.  
I train the model, and do evaluation and predicition , the result is consistent good, saying acc = 93%. 


But if i only do inference, i.e., loading the  saved model in outputs/,  the result is very low for the same test data. 
 i relaod througth below ways : 
 model = ClassificationModel('bert', 'outputs/', num_labels=30, args='outputs/model_args.json')
or  model = ClassificationModel('bert', 'outputs/', num_labels=len(map_cat_int.values()), args={})
    which i think it also read outputs/model_args.json

i also tried all the checkpoint models,  the acc are very low (acc<10%).   I debugged and can not fix it. 
Really appreciate anyone can tell me how this has been fixed in the end. Thanks very much .

**Comment by AbinayaM02** [comment]
Hi @ninasujit2016 : Could you please share the training_progress.csv file? It should be saved in the folder where you save your checkpoints.

**Comment by ninasujit2016** [comment]
sure, this is the training_progress.csv file i got it.  
global_step,mcc,train_loss,eval_loss,acc
931,0.8786946629258024,0.01930316723883152,0.5105949197534146,0.8869300911854103
1862,0.9143617513170853,0.00429063243791461,0.39385860651342686,0.9203647416413374
2000,0.9136495209305009,2.731369972229004,0.41321927979950185,0.9197568389057751
2793,0.9195604978114452,0.011454629711806774,0.3924517130470175,0.9252279635258359

**Comment by AbinayaM02** [comment]
From the output, it looks like the best model is from global_step 1862 (since the eval_loss increase at the next global_step). This will be the model that will get loaded when you load the model from best_model dir. Ideally, while inferencing, you should get 92% according the above output.  

I'm assuming the outputs folder is your best_model folder. If not, look for the best_model within the outputs/ folder. Use that as your model path in the below code.
model = ClassificationModel('bert', 'outputs/best_model', num_labels=30, args='outputs/best_model/model_args.json')

Try this and see if you're getting the right accuracy. If this doesn't work, then something else might be the problem.

**Comment by ninasujit2016** [comment]
Thanks Abinaya, Yes I tried all folders (all checkpoints too),  i tried 
model = ClassificationModel('bert', 'outputs', num_labels=30, args='outputs/model_args.json')
model = ClassificationModel('bert', 'outputs/best_model', num_labels=30, args='outputs/best_model/model_args.json')
the result by using ./outputs is slightly better than the result by using ./outputs/best_model/.  but both are quite low , acc <10%

**Comment by AbinayaM02** [comment]
Not sure what might be the problem. Do you mind sharing the code snippet?

**Comment by ninasujit2016** [comment]
model training and evaluation result are good , code see below 
------------------------------------------------------------------
        model_args = ClassificationArgs()
        model_args.early_stopping_metric_minimize = False
        model_args.early_stopping_metric = "acc"  # eval_loss
        model_args.evaluate_during_training = True
        #be default, the lowest eval_loss considerted to be the best model
        model_args.num_train_epochs = 3
        model_args.fp16 = False
        model_args.dataloader_num_workers = 0
        model_args.reprocess_input_data = True
        model_args.overwrite_output_dir = True
         model_args.gradient_accumulation_steps = 4
        model_args.train_batch_size = 2
        model_args.best_model_dir='./outputs/best_model/'
        model = ClassificationModel('bert', 'bert-base-cased', num_labels=len(map_cat_int.values()), args= model_args)
        #model_args = ClassificationArgs()
        # Train the model
          model.train_model(train_df,eval_df=eval_df,,acc=sklearn.metrics.accuracy_score)

        # save evaluaiton results
        result, model_outputs, wrong_predictions = model.eval_model(test_df, acc=sklearn.metrics.accuracy_score)

        # predict test data
        predictions, raw_outputs = model.predict(testsent)
---------------------------------------------------------------------------------
inference: 
# load model
model = ClassificationModel('bert', 'outputs', num_labels=len(map_cat_int.values()), args='outputs/model_args.json')
# prediction on the same test_df data used previously 
result, model_outputs, wrong_predictions = model.eval_model(test_df, acc=sklearn.metrics.accuracy_score)

**Comment by AbinayaM02** [comment]
Ideally, the accuracy should be same! Try saving the result and model_output as csv and check them manually on both the training and loading time. So far the test_df, in both the scenarios, ideally you should be getting both the results and model_outputs should be same.

**Comment by ninasujit2016** [comment]
thanks very much.  i think so and also expect the results are the same.  
This tool is easy to use and very useful if there is no requirement for inference.

**Comment by ninasujit2016** [comment]
BTW, my problem is not mulit-label, it is multiclass problem.

**Comment by thiagosantos1** [comment]
@ninasujit2016  What I have noticed is that when you give the number of labels and/or arguments, a new instance model is created and it doesn't load the previous one.

Have you tried :
- model = ClassificationModel('bert', 'outputs',  args='outputs/model_args.json')
OR
- model = ClassificationModel('bert', 'outputs', num_labels=30, args={ 'eval_batch_size': 32})

**Comment by ninasujit2016** [comment]
thanks very much for your time,  Abinaya and  thiagosantos . 
I found the reason why.   
It is due to the category mapping problem. Need to sort the label lists and do the mapping. 
After fixing the categories to id mapping, the acc is the same now.

**Comment by AbinayaM02** [comment]
Ah! I see. Glad you found it @ninasujit2016.
